composer/loggers/mosaicml_logger.py

```diff
@@ -0,0 +1,174 @@
+
+        self.sweep_id = sweep_id if sweep_id is not None else os.environ.get(_SWEEP_ID_ENV, None)
+        if self.sweep_id is None:
+            self.skip_logging = True
+            warnings.warn("No sweep_id provided to MosaicLoggerBackend. MosaicML logger will be a no-op.")
+
+        if creds_file:
+            with open(creds_file, 'r') as f:
+                os.environ[_MOSAICML_API_KEY_ENV] = str(f.read())
+
+        if os.environ.get(_MOSAICML_API_KEY_ENV, None) is None:
+            self.skip_logging = True
+            warnings.warn(
+                f"No api_key set for environment variable {_MOSAICML_API_KEY_ENV}. MosaicML logger will be a no-op.")
+
+        self.buffered_data = []
+        self.flush_every_n_batches = flush_every_n_batches
+        self.max_logs_in_buffer = max_logs_in_buffer
+
+        self.queue = Queue()
+        self.thread = Thread(target=self._listen_to_queue)
+
+    def _log_metric(self, epoch: int, step: int, log_level: LogLevel, data: TLogData):
+        del log_level  # unused
+
+        if self.skip_logging:
+            return
+
+        log_data = {
+            "job_id": self.job_id,
+            "sweep_id": self.sweep_id,
+            "epoch": epoch,
+            "step": step,
+            "data": data,
+        }
+        self.buffered_data.append(log_data)
+        if len(self.buffered_data) > self.max_logs_in_buffer:
+            self._flush_buffered_data()
+
+    def _training_start(self, state: State, logger: Logger) -> None:
+        del state, logger  # unused
+
+        if self.skip_logging:
+            return
+
+        log.info("Starting MosaicML logger thread.")
+
+        # Start the logging thread
+        self.thread.start()
+
+    def batch_end(self, state: State, logger: Logger):
+        del logger  # unused
+
+        if self.skip_logging:
+            return
+
+        if (state.step + 1) % self.flush_every_n_batches == 0:
+            self._flush_buffered_data()
+
+    def training_end(self, state: State, logger: Logger):
+        del state, logger  # unused
+
+        if self.skip_logging:
+            return
+
+        # Flush any remaining logs on training end
+        self._flush_buffered_data()
+
+        self.queue.put_nowait(_STOP_LOGGING_SIGNAL)
+        self.thread.join()
+
+        log.info(f"MosaicML logger thread has exited.")
+
+    def state_dict(self) -> StateDict:
+        # Storing these fields in the state dict to support run resuming in the future.
+        return {"job_id": self.job_id, "sweep_id": self.sweep_id}
+
+    def load_state_dict(self, state: StateDict) -> None:
+        self.job_id = state["job_id"]
+        self.sweep_id = state["sweep_id"]
+
+    def _flush_buffered_data(self):
+        if len(self.buffered_data) == 0:
+            return
+
+        data_to_write = self.buffered_data.copy()
+        self.buffered_data = []
+
+        self.queue.put_nowait(data_to_write)
+
+    def _listen_to_queue(self):
```

Comment:
Thoughts about having multiple listening threads for parallel uploads?

Reply:
Given that it's only flushing every n logs / every n batches and we have a reasonable timeout set, one thread is probably enough for now